src/basler_synced_cameras_node.cpp

```diff
@@ -0,0 +1,255 @@
+
+        bool camera_found = false;
+        while(ros::ok() && !camera_found)
+        {
+            tl_factory.EnumerateDevices(devices);
+            for (size_t i = 0; i < devices.size(); ++i)
+            {
+                if (devices[i].GetSerialNumber().c_str() == serial)
+                {
+                    ROS_INFO_STREAM("Found camera " << index << " with matching serial " << serial);
+                    cameras[index].Attach(tl_factory.CreateDevice(devices[i]));
+                    camera_found = true;
+                    break;
+                }
+            }
+            if (!camera_found)
+            {
+                ROS_INFO_STREAM("Failed to find camera " << index << " with serial " << serial);
+                ROS_INFO_STREAM( devices.size() << " other serials found:");
+                for (size_t i = 0; i < devices.size(); ++i)
+                {
+                    ROS_INFO_STREAM(devices[i].GetSerialNumber());
+                }
+                r.sleep();
+                ros::spinOnce();
+            }
+        }
+    }
+    return cameras;
+}
+
+// Wait for all cameras to be slaves and to have same master clock id
+void waitForPTPSlave(const std::vector<Pylon::CBaslerGigEInstantCamera>& cameras)
+{
+    ROS_INFO("Waiting for %zu cameras to be PTP slaves...", cameras.size());
+    ros::Rate r(0.5);
+    while (ros::ok())
+    {
+        int num_init = 0;
+        int num_master = 0;
+        int num_slave = 0;
+        int64_t master_clock_id = 0;
+        bool multiple_masters = false;
+        for (const auto& camera : cameras)
+        {
+            camera.GevIEEE1588DataSetLatch();
+            switch(camera.GevIEEE1588StatusLatched())
+            {
+            case Basler_GigECamera::GevIEEE1588StatusLatched_Master:
+            {
+                ++num_master;
+                break;
+            }
+            case Basler_GigECamera::GevIEEE1588StatusLatched_Slave:
+            {
+                int64_t master_id = camera.GevIEEE1588ParentClockId();
+                if (num_slave == 0)
+                {
+                    master_clock_id = master_id;
+                }
+                else if(master_id != master_clock_id)
+                {
+                    multiple_masters = true;
+                }
+                ++num_slave;
+                break;
+            }
+            default:
+            {
+                ++num_init;
+            }
+            }
+        }
+        if (num_slave == cameras.size() && !multiple_masters)
+        {
+            ROS_INFO_STREAM("All camera clocks are PTP slaves to master clock " << master_clock_id);
+            break;
+        }
+        else if(num_slave == cameras.size())
+        {
+            ROS_INFO("All camera clocks are PTP slaves, but multiple masters are present");
+        }
+        else
+        {
+            ROS_INFO("Camera PTP status: %d initializing, %d masters, %d slaves", num_init, num_master, num_slave);
```

Comment:
Can you throttle these outputs?  Maybe max one every 10s?  Otherwise if it takes a while to sync, that'll be a lot of messages!

You can also output it just the first time (or switch to debug).

options: http://wiki.ros.org/roscpp/Overview/Logging